classes/actions.php

```
@@ -147,7 +147,15 @@ public static function duplicate(array $modules, $sectionnumber = false): void {
         // sorted by their id:
         // Let order of mods in a section be mod1, mod2, mod3, mod4, mod5. If we duplicate mod2, mod4, the order afterwards will be
         // mod1, mod2, mod3, mod4, mod5, mod2(dup), mod4(dup).
+        $duplicatedmods = [];
+        $targetformat = course_get_format($courseid);
+        $sectionsrestricted = massactionutils::get_restricted_sections($courseid, $targetformat->get_format());
         foreach ($idsincourseorder as $cmid) {
+            $cm = $modinfo->get_cm($cmid);
+            // Not duplicated if the section is restricted.
+            if (in_array($cm->sectionnum, $sectionsrestricted)) {
+                continue;
```

Should we throw an exception of sectionrestricted here?